oauth2_http/java/com/google/auth/oauth2/JwtCredentials.java

```diff
@@ -0,0 +1,268 @@
+import com.google.api.client.util.Clock;
+import com.google.auth.Credentials;
+import com.google.auth.http.AuthHttpConstants;
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.io.Serializable;
+import java.net.URI;
+import java.security.GeneralSecurityException;
+import java.security.PrivateKey;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+public class JwtCredentials extends Credentials {
+  private static final String JWT_ACCESS_PREFIX = OAuth2Utils.BEARER_PREFIX;
+  private static final String JWT_INCOMPLETE_ERROR_MESSAGE = "JWT claims must contain audience, "
+      + "issuer, and subject.";
+
+  // byte[] is serializable, so the lock variable can be final
+  private final Object lock = new byte[0];
+  private final PrivateKey privateKey;
+  private final String privateKeyId;
+  private final Claims claims;
+  private final Long lifeSpanSeconds;
+  @VisibleForTesting
+  transient Clock clock;
+
+  private transient String jwt;
+  private transient Long expiry;
+
+  JwtCredentials(Builder builder) {
+    this.privateKey = Preconditions.checkNotNull(builder.getPrivateKey());
+    this.privateKeyId = Preconditions.checkNotNull(builder.getPrivateKeyId());
+    this.claims = Preconditions.checkNotNull(builder.getClaims());
+    Preconditions.checkState(claims.isComplete(), JWT_INCOMPLETE_ERROR_MESSAGE);
+    this.lifeSpanSeconds = Preconditions.checkNotNull(builder.getLifeSpanSeconds());
+    this.clock = Preconditions.checkNotNull(builder.getClock());
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  @Override
+  public void refresh() throws IOException {
+    synchronized (lock) {
+      JsonWebSignature.Header header = new JsonWebSignature.Header();
+      header.setAlgorithm("RS256");
+      header.setType("JWT");
+      header.setKeyId(privateKeyId);
+
+      JsonWebToken.Payload payload = new JsonWebToken.Payload();
+      long currentTime = clock.currentTimeMillis();
+      payload.setAudience(claims.getAudience());
+      payload.setIssuer(claims.getIssuer());
+      payload.setSubject(claims.getSubject());
+      payload.setIssuedAtTimeSeconds(currentTime / 1000);
+      expiry = currentTime / 1000 + lifeSpanSeconds;
+      payload.setExpirationTimeSeconds(expiry);
+
+      JsonFactory jsonFactory = OAuth2Utils.JSON_FACTORY;
+
+      try {
+        jwt = JsonWebSignature.signUsingRsaSha256(privateKey, jsonFactory, header, payload);
+      } catch (GeneralSecurityException e) {
+        throw new IOException("Error signing service account JWT access header with private key.",
+            e);
+      }
+    }
+  }
+
+  private boolean shouldRefresh() {
+    return expiry == null || getClock().currentTimeMillis() / 1000 > expiry;
```

Comment:
This should incorporate the standard clock skew accommodation: https://github.com/googleapis/google-auth-library-python/blob/master/google/auth/_helpers.py#L25